sql/src/main/java/io/crate/execution/dml/upsert/TransportShardUpsertAction.java

```diff
@@ -309,53 +306,108 @@ private IndexItemResponse indexItem(ShardUpsertRequest request,
+            // when return values are requested
+            if (source == null) {
+                source = JsonXContent.jsonXContent.createParser(
+                    NamedXContentRegistry.EMPTY,
+                    DeprecationHandler.THROW_UNSUPPORTED_OPERATION,
+                    BytesReference.toBytes(rawSource)).map();
             }
+            returnvalues = returnGen.generateReturnValues(
+                // return -1 as docId, the docId can only be retrieved by fetching the inserted document again, which
+                // we want to avoid. The docId is anyway just valid with the lifetime of a searcher and can change afterwards.
+                new Doc(-1,
+                        indexShard.shardId().getIndexName(),
+                        item.id(),
+                        indexResult.getVersion(),
+                        indexResult.getSeqNo(),
+                        primaryTerm,
+                        source,
+                    () -> ""
+                )
+            );
+        }
+        return new IndexItemResponse(indexResult.getTranslogLocation(), returnvalues);
+    }
 
-            item.source(BytesReference.bytes(XContentFactory.jsonBuilder().map(updatedSource)));
-            seqNo = item.seqNo();
-            primaryTerm = item.primaryTerm();
-            version = Versions.MATCH_ANY;
+    protected IndexItemResponse update(ShardUpsertRequest.Item item,
+                                       IndexShard indexShard,
+                                       boolean isRetry,
+                                       @Nullable ReturnValueGen returnGen,
+                                       @Nullable UpdateSourceGen updateSourceGen) throws Exception {
+        assert updateSourceGen != null : "UpdateSourceGen must not be null";
+        Doc fetchedDoc = getDocument(indexShard, item.id(), item.version(), item.seqNo(), item.primaryTerm());
+        Map<String, Object> source = updateSourceGen.generateSource(
+            fetchedDoc,
+            item.updateAssignments(),
+            item.insertValues()
+        );
+        BytesReference rawSource = BytesReference.bytes(XContentFactory.jsonBuilder().map(source));
+        item.source(rawSource);
+        long seqNo = item.seqNo();
+        long primaryTerm = item.primaryTerm();
+        long version = Versions.MATCH_ANY;
+
+        Engine.IndexResult indexResult = index(item, indexShard, isRetry, seqNo, primaryTerm, version);
+        Object[] returnvalues = null;
+        if (returnGen != null) {
+            returnvalues = returnGen.generateReturnValues(
+                new Doc(fetchedDoc.docId(),
+                        fetchedDoc.getIndex(),
+                        fetchedDoc.getId(),
+                        indexResult.getVersion(),
+                        indexResult.getSeqNo(),
+                        indexResult.getTerm(),
+                        source,
+                    () -> "")
```

Comment:
Checkstyle lambda oddity

Reply:
See my comment further above, if each arg is on a newline checkstyle passes.